### ISSUE TITLE
Issue: ActiveRecord PG - Issue deleting records (in execute_and_clear & exec_delete)

### ISSUE THREAD

**Issue by nathan-appere** [rails/rails#51460]
### Steps to reproduce

I sadly cannot reproduce locally or on preview envs, it only happens in our production environment 😢 

### Expected behavior

AR lets you delete a record.

### Actual behavior

On some delete actions triggered by a `record.delete` we get either:

```
undefined method `clear' for nil:NilClass (NoMethodError)
activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql_adapter.rb in ensure in execute_and_clear
```

or

```
undefined method `cmd_tuples' for []:Array (NoMethodError)
activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql/database_statements.rb in block in exec_delete
```

### System configuration
**Rails version**: `7.1.3.2`

**Ruby version**: `3.2.3`

**pg version**:  `1.5.6`

**Comment by byroot** [comment]
Please at least include the full backtraces.

**Comment by nathan-appere** [comment]
```
NoMethodError
undefined method `cmd_tuples' for []:Array (NoMethodError)
          execute_and_clear(sql, name, binds) { |result| result.cmd_tuples }

activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql/database_statements.rb in block in exec_delete
activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql_adapter.rb in execute_and_clear at line 877
activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql/database_statements.rb in exec_delete at line 77
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/database_statements.rb in delete at line 208
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/query_cache.rb in delete at line 25
activerecord (7.1.3.2) lib/active_record/persistence.rb in _delete_record at line 624
activerecord (7.1.3.2) lib/active_record/persistence.rb in _delete_row at line 1198
activerecord (7.1.3.2) lib/active_record/persistence.rb in destroy_row at line 1194
activerecord (7.1.3.2) lib/active_record/counter_cache.rb in destroy_row at line 197
activerecord (7.1.3.2) lib/active_record/locking/optimistic.rb in destroy_row at line 125
activerecord (7.1.3.2) lib/active_record/persistence.rb in destroy at line 783
activerecord (7.1.3.2) lib/active_record/callbacks.rb in block in destroy at line 423
activesupport (7.1.3.2) lib/active_support/callbacks.rb in run_callbacks at line 110
activesupport (7.1.3.2) lib/active_support/callbacks.rb in _run_destroy_callbacks at line 952
activerecord (7.1.3.2) lib/active_record/callbacks.rb in destroy at line 423
activerecord (7.1.3.2) lib/active_record/transactions.rb in block in destroy at line 305
activerecord (7.1.3.2) lib/active_record/transactions.rb in block in with_transaction_returning_status at line 365
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/transaction.rb in block in within_new_transaction at line 535
activesupport (7.1.3.2) lib/active_support/concurrency/null_lock.rb in synchronize at line 9
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/transaction.rb in within_new_transaction at line 532
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/database_statements.rb in transaction at line 344
activerecord (7.1.3.2) lib/active_record/transactions.rb in with_transaction_returning_status at line 361
activerecord (7.1.3.2) lib/active_record/transactions.rb in destroy at line 305
```

and

```
NoMethodError
undefined method `clear' for nil:NilClass (NoMethodError)
            result.clear

activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql_adapter.rb in execute_and_clear at line 879
activerecord (7.1.3.2) lib/active_record/connection_adapters/postgresql/database_statements.rb in exec_delete at line 77
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/database_statements.rb in delete at line 208
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/query_cache.rb in delete at line 25
activerecord (7.1.3.2) lib/active_record/persistence.rb in _delete_record at line 624
activerecord (7.1.3.2) lib/active_record/persistence.rb in _delete_row at line 1198
activerecord (7.1.3.2) lib/active_record/persistence.rb in destroy_row at line 1194
activerecord (7.1.3.2) lib/active_record/counter_cache.rb in destroy_row at line 197
activerecord (7.1.3.2) lib/active_record/locking/optimistic.rb in destroy_row at line 125
activerecord (7.1.3.2) lib/active_record/persistence.rb in destroy at line 783
activerecord (7.1.3.2) lib/active_record/callbacks.rb in block in destroy at line 423
activesupport (7.1.3.2) lib/active_support/callbacks.rb in run_callbacks at line 110
activesupport (7.1.3.2) lib/active_support/callbacks.rb in _run_destroy_callbacks at line 952
activerecord (7.1.3.2) lib/active_record/callbacks.rb in destroy at line 423
activerecord (7.1.3.2) lib/active_record/transactions.rb in block in destroy at line 305
activerecord (7.1.3.2) lib/active_record/transactions.rb in block in with_transaction_returning_status at line 365
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/transaction.rb in block in within_new_transaction at line 535
activesupport (7.1.3.2) lib/active_support/concurrency/null_lock.rb in synchronize at line 9
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/transaction.rb in within_new_transaction at line 532
activerecord (7.1.3.2) lib/active_record/connection_adapters/abstract/database_statements.rb in transaction at line 344
activerecord (7.1.3.2) lib/active_record/transactions.rb in with_transaction_returning_status at line 361
activerecord (7.1.3.2) lib/active_record/transactions.rb in destroy at line 305
```

**Comment by casperisfine** [comment]
` result.clear` at https://github.com/rails/rails/blob/v7.1.3.2/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L879

```ruby
          if !prepare || without_prepared_statement?(binds)
            result = exec_no_cache(sql, name, binds, async: async, allow_retry: allow_retry, materialize_transactions: materialize_transactions)
          else
            result = exec_cache(sql, name, binds, async: async, allow_retry: allow_retry, materialize_transactions: materialize_transactions)
          end
          begin
            ret = yield result
          ensure
            result.clear
          end
          ret
```

This suggest `exec_cache` or `exec_no_cache` returned `nil`, which shouldn't happen.

**Comment by casperisfine** [comment]
> it only happens in our production environment

Any patterns on how it happen? e.g. fully random? Or some spikes around a certain time across all processes, or only in one process, etc?

**Comment by nathan-appere** [comment]
@casperisfine figured it out, this was somehow caused by a missing `dependent:  :destroy` on a model, no idea how that materialized as nil `result` though 🤔

**Comment by byroot** [comment]
Yeah, that's extremely weird. If you ever come up with a repro, please share it. Because I suspect there is more to it.

**Comment by trevorturk** [comment]
I'm not sure if this is related, but I wanted to mention that I encountered the same `result.clear` error where `exec_cache` and `exec_no_cache` (I tested both) were `nil` (which shouldn't happen) when I upgraded from Rails 7.0.x to 7.1.x in an app running Falcon where I hadn't switched `isolation_state` to `:fiber`. Once I switched the isolation state to fiber, the error disappeared. (There were a few other strange errors like this which all disappeared, but this was the most common one by far, and not just triggered by deleting records.)

I'm curious if @nathan-appere might be using Falcon and/or Fibers? I'm sorry if this isn't related or is misdirecting from the actual issue, but I thought it worth mentioning. Thanks!
